### PR TITLE
Add event-selection task in runtest

### PR DESCRIPTION
**Pull request by fgrosa**
Automatically enable event selection and timestamp tasks if track-index creator enabled. Necessary for the modifications in this O2 PR https://github.com/AliceO2Group/AliceO2/pull/6016

### CODE REVIEW

**Comment by vkucera** [comment]
Please solve conflicts and adapt to the new workflow database.

**Comment by fgrosa** [comment]
> Please solve conflicts and adapt to the new workflow database.

done in https://github.com/AliceO2Group/Run3Analysisvalidation/pull/192/commits/9b5e0d89ed6005dc1037839d211e2cfec1e74cbe

**Comment by vkucera** [comment]
I fixed some details.
Doesn't `dpl-config_run5_oniaX.json` also need changes?

**Comment by vkucera** [comment]
Also please keep the order of configurables the same in all jsons.

**Comment by fgrosa** [comment]
> I fixed some details.
> Doesn't `dpl-config_run5_oniaX.json` also need changes?

Thanks! I did not realise that a `json` was added, I updated it now in https://github.com/AliceO2Group/Run3Analysisvalidation/pull/192/commits/5e9210b3031d48aa85fa3a41a3eac3ce4d31299b. I changed the position of `"ptmintrack_2prong": "1.0",` to uniform it with the other two `json`

**Comment by vkucera** [comment]
Hi @fgrosa , I am not sure the "Make explicit the activation of the event selection" commit is the right way to enable this.
I would prefer to keep the `make_command_o2.py` script quite general and keep the specific options in the database.
Adding workflow-specific switches in the script will eventually lead to a similar mess as was in `config_tasks.sh` that I tried to clean up by making this script.

**Comment by fgrosa** [comment]
> Hi @fgrosa , I am not sure the "Make explicit the activation of the event selection" commit is the right way to enable this.
> I would prefer to keep the `make_command_o2.py` script quite general and keep the specific options in the database.
> Adding workflow-specific switches in the script will eventually lead to a similar mess as was in `config_tasks.sh` that I tried to clean up by making this script.

Hi @vkucera the idea behind this implementation was that by enabling `DOO2_EVSEL` in the `config_task.sh` one has it also automatically activated in the `HFTrackIndexSkim.cxx` task and hence should not care about consistency between enabled tasks and arguments in the database, but if you prefer that I add an argument to the `json` instead there is no problem

**Comment by fgrosa** [comment]
Hi @vkucera, I found a solution that does not required to modify the `make_command_o2.py`, could you check if it looks fine for you?

**Comment by vkucera** [comment]
> Hi @vkucera, I found a solution that does not required to modify the `make_command_o2.py`, could you check if it looks fine for you?

Thanks @jgrosseo . This is indeed what I suggested. However, it doesn't work if `o2-analysis-hf-track-index-skims-creator` is not the only dependency.
You have to change the dependency fields containing `o2-analysis-hf-track-index-skims-creator` to be all in the form of a list with one workflow per line:
```yaml
  dependencies:
    - o2-analysis-hf-track-index-skims-creator
```
Then it should work for any workflow.

**Comment by fgrosa** [comment]
> > Hi @vkucera, I found a solution that does not required to modify the `make_command_o2.py`, could you check if it looks fine for you?
> 
> Thanks @fgrosa . This is indeed what I suggested. However, it doesn't work if `o2-analysis-hf-track-index-skims-creator` is not the only dependency.
> You have to change the dependency fields containing `o2-analysis-hf-track-index-skims-creator` to be all in the form of a list with one workflow per line:
> 
> ```yaml
>   dependencies:
>     - o2-analysis-hf-track-index-skims-creator
> ```
> 
> Then it should work for any workflow.

Done! :thumbsup: I also implemented the feature that allows us to run cascades together with other HF workflows as discussed offline, let me know if it is ok for you

File: codeHF/config_tasks.sh [hunk 1]
```diff
@@ -141,6 +149,12 @@ function AdjustJson {
 # Generate the O2 script containing the full workflow specification.
 function MakeScriptO2 {
   WORKFLOWS=""
+  # Cascade reconstruction
+  CASCADESUFFIX=""
+  [ $DOO2_CASC -eq 1 ] && CASCADESUFFIX+="_v0"
+  # Event selection
+  EVSELSUFFIX=""
+  [ $DOO2_EVSEL -eq 1 ] && WORKFLOWS+=" o2-analysis-timestamp o2-analysis-event-selection" && EVSELSUFFIX+="-evsel"
```
Comment (vkucera) [inline codeHF/config_tasks.sh:157]: * Better to use ordered names, matching the switches, like `SUFFIX_CASC`, `SUFFIX_EVSEL`.
* You can do a single line: `[ $DOO2_CASC -eq 1 ] && CASCADESUFFIX="_v0" || CASCADESUFFIX=""`.
```bash
[ $DOO2_EVSEL -eq 1 ] && WORKFLOWS+=" o2-analysis-timestamp o2-analysis-event-selection" && EVSELSUFFIX+="-evsel"
```

is not the right syntax. Instead use:

```bash
[ $DOO2_EVSEL -eq 1 ] && { WORKFLOWS+=" o2-analysis-timestamp o2-analysis-event-selection"; EVSELSUFFIX="-evsel"; }
```
* For consistency, I would replace `_v0` with `-v0` everywhere.

File: codeHF/config_tasks.sh [hunk 2]
```diff
@@ -150,8 +164,7 @@ function MakeScriptO2 {
   [ $DOO2_PID_TOF -eq 1 ] && WORKFLOWS+=" o2-analysis-pid-tof-full"
   [ $DOO2_PID_TOF_QA -eq 1 ] && WORKFLOWS+=" o2-analysis-pid-tof-qa-mc"
   # Vertexing
-  [ $DOO2_SKIM -eq 1 ] && WORKFLOWS+=" o2-analysis-hf-track-index-skims-creator"
-  [ $DOO2_SKIM_V0 -eq 1 ] && WORKFLOWS+=" o2-analysis-hf-track-index-skims-creator_v0"
+  [ $DOO2_SKIM -eq 1 ] && WORKFLOWS+=" o2-analysis-hf-track-index-skims-creator${EVSELSUFFIX}${CASCADESUFFIX}"
```
Comment (vkucera) [inline codeHF/config_tasks.sh:167]: Maybe better to define a variable here like `WF_SKIM="o2-analysis-hf-track-index-skims-creator${EVSELSUFFIX}${CASCADESUFFIX}"`
and use it later with `ReplaceString`

File: codeHF/config_tasks.sh [hunk 3]
```diff
@@ -182,6 +195,17 @@ function MakeScriptO2 {
   [ $SAVETREES -eq 1 ] && OPT_MAKECMD+=" -t"
   [ $MAKE_GRAPH -eq 1 ] && OPT_MAKECMD+=" -g"
 
+  # Make a copy of the default workflow database file before modifying it
+  if [[ $DOO2_EVSEL -eq 1 || $DOO2_CASC -eq 1 ]]; then
+    DATABASE_O2_EDIT=".${DATABASE_O2}"
```
Comment (vkucera) [inline codeHF/config_tasks.sh:200]: Define empty `DATABASE_O2_EDIT` before `if`.
Why `".${DATABASE_O2}"` instead of `"${DATABASE_O2/.yml/_edit.yml}"`?

**Review by vkucera — CHANGES_REQUESTED** [review]

File: codeHF/config_tasks.sh [hunk 4]
```diff
@@ -74,6 +75,12 @@ SAVETREES=0         # Save O2 tables to trees.
 USEO2VERTEXER=0     # Use the O2 vertexer in AliPhysics.
 DORATIO=0           # Plot histogram ratios in comparison.
 
+# enable cascade reconstruction by default in case of Λc → K0S p tasks
+DOO2_CASC=0
+if [[ $DOO2_CAND_CASC -eq 1 || $DOO2_SEL_LCK0SP -eq 1 || $DOO2_TASK_LCK0SP -eq 1 || $APPLYCUTS_LCK0SP -eq 1 ]]; then
+  DOO2_CASC=1
```
[the remark this answers is not in the document]
Comment (fgrosa) [reply codeHF/config_tasks.sh]: Indeed it doesn't, I had overlooked it

File: codeHF/config_tasks.sh (hunk 1 above)
Comment (fgrosa) [reply codeHF/config_tasks.sh]: Done!

File: codeHF/config_tasks.sh (hunk 2 above)
Comment (fgrosa) [reply codeHF/config_tasks.sh]: Done!

File: codeHF/config_tasks.sh (hunk 3 above)
Comment (fgrosa) [reply codeHF/config_tasks.sh]: Fixed

File: codeHF/config_tasks.sh [hunk 5]
```diff
@@ -141,6 +149,9 @@ function AdjustJson {
 # Generate the O2 script containing the full workflow specification.
 function MakeScriptO2 {
   WORKFLOWS=""
+  # Cascade reconstruction
+  [ $DOO2_CASC -eq 1 ] && SUFFIX_CASC="-v0" || SUFFIX_CASC=""  # Event selection
+  [ $DOO2_EVSEL -eq 1 ] && { WORKFLOWS+=" o2-analysis-timestamp o2-analysis-event-selection"; SUFFIX_EVSEL="-evsel"; } || SUFFIX_EVSEL=""
```
Comment (vkucera) [inline codeHF/config_tasks.sh:154]: Fix the position of `# Event selection`.

File: codeHF/config_tasks.sh [hunk 6]
```diff
@@ -28,6 +28,8 @@ DATABASE_O2="workflows.yml"
 MAKE_GRAPH=0        # Make topology graph.
 
 # Activation of O2 workflows
+# Event selection
+DOO2_EVSEL=1        # event-selection and timestamp
```
Comment (vkucera) [inline codeHF/config_tasks.sh:32]: Should be `0` by default.

File: codeHF/config_tasks.sh [hunk 7]
```diff
@@ -57,7 +58,7 @@ DOO2_TASK_XIC=0     # hf-task-xic
 DOO2_TASK_JPSI=0    # hf-task-jpsi
 DOO2_TASK_BPLUS=0   # hf-task-bplus
 DOO2_TASK_X=0       # hf-task-x
-DOO2_TASK_LCK0SP=0  # hf-task-lc-tok0sp
+DOO2_TASK_LCK0SP=1  # hf-task-lc-tok0sp
```
Comment (vkucera) [inline codeHF/config_tasks.sh:61]: Should be `0` by default.

File: codeHF/workflows.yml [hunk 8]
```diff
@@ -5,25 +5,52 @@ options:
...
     dependencies: o2-analysis-lambdakzerobuilder
     options: "--do-LcK0Sp"
     tables: [HFSELTRACK, HFTRACKIDXCASC]
 
+  o2-analysis-hf-track-index-skims-creator-evsel-v0:
+    <<: *skim_creator
+    activate: false
+    dependencies: [o2-analysis-timestamp, o2-analysis-event-selection, o2-analysis-lambdakzerobuilder]
+    options: ["--doEvSel", "--do-LcK0Sp"]
+    tables: [HFSELCOL, HFSELTRACK, HFTRACKIDXCASC]
```
Comment (vkucera) [inline codeHF/workflows.yml:46]: Since the `-v0` versions can run with 2/3-prong tasks now, they should include also the related tables.

**Review by vkucera — CHANGES_REQUESTED** [review]
I think it's almost there. :-)

File: codeHF/config_tasks.sh (hunk 5 above)
Comment (fgrosa) [reply codeHF/config_tasks.sh]: Done

File: codeHF/config_tasks.sh [hunk 9]
```diff
@@ -182,6 +193,18 @@ function MakeScriptO2 {
...
+  # Make a copy of the default workflow database file before modifying it
+  DATABASE_O2_EDIT=""
+  if [[ $DOO2_EVSEL -eq 1 || $DOO2_CASC -eq 1 ]]; then
+    DATABASE_O2_EDIT="${DATABASE_O2/.yml/_edit.yml}"
+
+    cp "$DATABASE_O2" "$DATABASE_O2_EDIT" || ErrExit "Failed to cp $DATABASE_O2 $DATABASE_O2_EDIT."
+    DATABASE_O2="$DATABASE_O2_EDIT"
+
+    # Adjust workflow database in case of event selection enabled
+    ReplaceString "- o2-analysis-hf-track-index-skims-creator" "- ${WF_SKIM}" "$DATABASE_O2" || ErrExit "Failed to edit $DATABASE_O2."
```
[the remark this answers is not in the document]
Comment (fgrosa) [reply codeHF/config_tasks.sh]: Done!

File: codeHF/config_tasks.sh (hunk 6 above)
Comment (fgrosa) [reply codeHF/config_tasks.sh]: Indeed this was a leftover of the tests 😄 fixed!

File: codeHF/config_tasks.sh (hunk 7 above)
Comment (fgrosa) [reply codeHF/config_tasks.sh]: Same as above

File: codeHF/config_tasks.sh [hunk 10]
```diff
@@ -74,6 +75,12 @@ SAVETREES=0         # Save O2 tables to trees.
 USEO2VERTEXER=0     # Use the O2 vertexer in AliPhysics.
 DORATIO=0           # Plot histogram ratios in comparison.
 
+# enable cascade reconstruction by default in case of Λc → K0S p tasks
+DOO2_CASC=0
+if [[ $DOO2_CAND_CASC -eq 1 || $DOO2_SEL_LCK0SP -eq 1 || $DOO2_TASK_LCK0SP -eq 1 ]]; then
+  DOO2_CASC=1
+fi
```
[the remark this answers is not in the document]
Comment (fgrosa) [reply codeHF/config_tasks.sh]: Done!

File: codeHF/workflows.yml (hunk 8 above)
Comment (fgrosa) [reply codeHF/workflows.yml]: Right, I forgot that, now it is fixed

**Review by vkucera — APPROVED** [review]

**Pull request merged**